Storage/REST-API.md

```diff
@@ -0,0 +1,185 @@
+
+  For `PATCH` requests, the key of the record being updated is taken from the URL. The user is expected to either not send in the key as part of the payload, or ensure the key contained in the payload matches the key contained in the URL.
+  
+  `PATCH` will fail if attempted to modify a non-existing resource, resulting in `404 (Not Found)` error code.
+  
+  Also, `PATCH` performs conflict resolution. The URL being `PATCH`-ed must contain a revision / timestamp of the resource. The exact semantics of the revision / timestamp is not specified by this RFC, but the top-level `"url"` field returned by the `GET` request is guaranteed to have it set.
+
+  An attempt to `PATCH` to a URL that points to the version of the resource other than the most recent one will result in no data mutation and a `409 Conflict` error code. The user is then expected to re-run the `GET` request, followed by the re-run `PATCH` request onto the URL of the most recent version of the resource returned by the `GET` request. Confirming that the modifications that occurred in the meantime do not conflict with the new payload being sent in the body of the `PATCH` request is then the responsibility of the API user.
+
+* **PUT**: Create a new record for a resource, or update all fields of an existing resource.
+
+  Attempting to `PUT` to a URL that has an incomplete view of a record would result in a `405 Method Not Allowed` error.
+  
+  In the `&format=full` view, `PUT` is equivalent to `PATCH`. 
+  
+  `PUT` behaves the same way as `PATCH` when it comes to avoiding merge conflicts.
+
+  When legal, the only difference between `PUT` and `PATCH` is that `PUT` will create a new resource if it did not exist before the call. A `201 Created` response code will then be returned. Use `PATCH` if this behavior is undesirable.
+
+  `TODO: Location header https://tools.ietf.org/html/rfc7231#section-4.3.3`
+
+* **POST**: Create a new record and assign it to some resourse.
+
+  `POST` will create a new resource. The URL for POST must not contain the key(s) for the record being posted, and the payload body should not contain the key(s) either.
+
+  The generation of the new key is user-defined. Three common strategies are: 1) random, 2) the [salted] hash of the payload, 3) the value of certain passed in field or fields.
+  
+  In the latter two, as well as other, cases, a `POST` request may result in an attempt to overwrite an already existing resource. By default, `POST` will not overwrite, but return a `409 Conflict` error code, with the top-level `"url"` field of the returned JSON object containing the URL for the resource that was about to be overwritten.
+
+  The `&overwrite=1` / `&overwrite=true` URL parameter can be set to indicate that the `POST` request should complete successfully even if it would result in overwriting existing data.
```

Comment:
There's a corner case: doing a `POST` without specifying the resource, of an object that contains enough data to imply what the resource should be, and that resource already existing.

For example, consider the DB where the server-side logic knows the primary key for users are e-mails. The user does not have to know it, and they do a POST. If the body of this POST request contains an object with an e-mail that already presents in the DB, a point can be made to have the default behavior not overwrite the existing record.

A PUT immediately following a POST that returned an error and a "canonical" URL for the resource to be updated is indeed a solution. It's extra logic on the client side though, hence we thought an `&overwrite=true` flag could be useful.